Readme.md
## snakemakelib ##

<table>
<tr>
<td>master</td>	
<td><img src="https://travis-ci.org/percyfal/snakemakelib.svg?branch=master" alt="master" /></td>
<td><img src="https://coveralls.io/repos/percyfal/snakemakelib/badge.svg?branch=master" alt="coverage master" /></td>
</tr>
<tr>
<td>develop</td>
<td><img src="https://travis-ci.org/percyfal/snakemakelib.svg?branch=develop" alt="develop" /></td>
<td><img src="https://coveralls.io/repos/percyfal/snakemakelib/badge.svg?branch=develop" alt="coverage develop" /></td>
</tr>
</table>

## About ##

Snakemake library for all kinds of tasks, with a focus on
bioinformatics and next-generation sequencing. See
[the documentation](http://snakemakelib.readthedocs.org/en/latest/index.html)
for more information.

## Installation ##

snakemakelib is still not on PyPI but you can still install from
github using pip in several ways. I recommend using the
[user installation scheme](https://docs.python.org/3.4/install/index.html#inst-alt-install-user)
by making use of the *--user* flag.

By default, user site-packages are placed in **~/.local**. If you
choose to install in a different location, you need to set the
environment variable
[PYTHONUSERBASE](https://docs.python.org/3.4/using/cmdline.html#envvar-PYTHONUSERBASE)
to e.g. *$HOME/lib/python*. Then, with this setting, set
[PYTHONPATH](https://docs.python.org/3.4/using/cmdline.html#envvar-PYTHONPATH)
to *$HOME/lib/python/lib/python3.4/site-packages* to access the
package. The examples below show the user installation scheme.

Alternatively, install the package into a
[virtual environment](http://docs.python-guide.org/en/latest/dev/virtualenvs/).

### Installing directly from github ###

You can install as an
[editable install](https://pip.pypa.io/en/latest/reference/pip_install.html#editable-installs)
by invoking

	pip3 install -e git+https://github.com/percyfal/snakemakelib.git@master#egg=snakemakelib --user

### Cloning and installing ###

The previous command currently fails to install
[bokehutils](https://github.com/percyfal/bokehutils) (see
[issue 16](https://github.com/percyfal/snakemakelib/issues/16)). As a
workaround, you can manually clone the repo

	git clone git@github.com:percyfal/snakemakelib.git

and invoke (standing in the cloned repo)

	pip3 install -r requirements.txt --user -e .

## Quickstart ##

Probably the best way to get started is to run a real example. See the
[repaper](<https://github.com/percyfal/repaper>) repository for
workflows that should work out of the box.


ChangeLog
Changelog
=========

0.1-alpha.15 (2015-09-29)
-------------------------

- Add final support for bokehutils-0.1.6. [Per Unneberg]

  Fix #46

- Minor changes to ATAC-seq. [Per Unneberg]

  - make picard summary pick up unique files
  - add ucsc to ATAC-seq workflow

- Use log2 values for pca plots. [Per Unneberg]

- Pickle pca output, minor changes to templates. [Per Unneberg]

- Add pca plots and templates. [Per Unneberg]

  Substantial additions including:

  - code to do pca and plot the results
  - template additions
  - minor fixes to the ordering on x-axes

- Minor fixes. [Per Unneberg]

- Add and update pca rule. [Per Unneberg]

- Add release notes for 0.1-alpha.15. [Per Unneberg]

- Convert numerical sample ids to strings. [Per Unneberg]

  Fixes #45

- Update dependency to bokehutils-0.1.6. [Per Unneberg]

  - update templates to use bokeh_css and bokeh_js

- Remove dirty tags from doc build. [Per Unneberg]

  Fixes #47

0.1-alpha.14 (2015-09-16)
-------------------------

- Update dependency to snakemake 3.4.2. [Per Unneberg]

- Disable limit in find_files. [Per Unneberg]

  Disable limit in find_files if limit key missing in regexp

0.1-alpha.13 (2015-09-10)
-------------------------

- Hotfix: fix #40. [Per Unneberg]

  scrnaseq merge rule did symlink on temporary dependency

0.1-alpha.12 (2015-09-08)
-------------------------

- Make correct rnaseq annotations. [Per Unneberg]

  Fix #39

0.1-alpha.11 (2015-09-07)
-------------------------

- Add release notes. [Per Unneberg]

- Update rules to work standalone. [Per Unneberg]

- Move data and include in setup. [Per Unneberg]

  Add db lambda functions and update docs

  Documentation update

  Documentation update

- Documentation updates. [Per Unneberg]

- Minor tweaks to rules and workflow. [Per Unneberg]

- Update navbar and add sidebar toc. [Per Unneberg]

- Ditto. [Per Unneberg]

- Updates to docs. [Per Unneberg]

- Documentation updates. [Per Unneberg]

- Minor fixes to rules. [Per Unneberg]

- Convert bamtools to atomic and fix rules. [Per Unneberg]

  Convert bamtools to atomic rules. Fix some bugs in rules.

- Fixes for scrnaseq workflow. [Per Unneberg]

  - add reporting rules for rseqc, rsem, star
  - update cloudbiolinux annotation gtf rule

- Cleanup python2 statements. [Per Unneberg]

  Statements have been pruned of " && "

- Simplify bowtie2_align. [Per Unneberg]

  Can take gzipped input

- Update config loading. [Per Unneberg]

  - change sml_path and friends to variables (fixes #38)
  - remove trailing refs to samtools_cfg and the likes (fixes #37)

0.1-alpha.10 (2015-09-01)
-------------------------

- Add snakemakelib.workflow to setup. [Per Unneberg]

  Fixes #34

0.1-alpha.9 (2015-08-31)
------------------------

- Update aligner config in atacseq. [Per Unneberg]

  Fixes #23

0.1-alpha.8 (2015-08-30)
------------------------

- Add release notes. [Per Unneberg]

  cdb40d0 fixes #21

- Add readthedocs test based on file name. [Per Unneberg]

- Add error message. [Per Unneberg]

- Add socket to get hostname. [Per Unneberg]

- Remove matplotlib dependency. [Per Unneberg]

0.1-alpha.7 (2015-08-29)
------------------------

- Add release notes. [Per Unneberg]

- Remove trailing diff line. [Per Unneberg]

- Add atomic_rules. [Per Unneberg]

  Fixes #12

  Fix various inconsistensies in workflows

- Add angsd and tuxedo atomic rules. [Per Unneberg]

- Add misc atomic rules. [Per Unneberg]

- Add vcftools. [Per Unneberg]

- Add atomic variation rules. [Per Unneberg]

  Add rules for annovar, snpeff, htslib, vcftools, vcf

- Add rsem atomic rules. [Per Unneberg]

- Add rpkmforgenes atomic rules. [Per Unneberg]

- Delete sequenceprocessing and convert to single applications. [Per
  Unneberg]

- Add cutadapt atomic rules. [Per Unneberg]

- Add rseqc atomic rules. [Per Unneberg]

- Add picard atomic rules. [Per Unneberg]

- Add gatk atomic rules. [Per Unneberg]

- Add and update atomic rules. [Per Unneberg]

- Add atomic rules. [Per Unneberg]

  Start converting to atomic rules

- Initial support of atomic rules. [Per Unneberg]

  * setup.py: distinguish between multiple rules and atomic rules with '.rule' and '.rules' suffixes
  * snakemakelib/config.py: don't require config dict is a BaseConfig instance
  * snakemakelib/rules/bio/ngs/align/bwa/bwa.rules: default settings for bwa
  * snakemakelib/rules/bio/ngs/align/bwa/bwa_index.rule: atomic bwa index rule
  * snakemakelib/rules/bio/ngs/align/bwa/bwa_link_ref.rule: atomic bwa link ref rule
  * snakemakelib/rules/bio/ngs/align/bwa/bwa_mem.rule: atomic bwa mem rule
  * snakemakelib/rules/bio/ngs/settings.rules: use tentative snakemake configdefault keyword
  * snakemakelib/rules/bio/ngs/tools/samtools.rules: ditto
  * snakemakelib/rules/settings.rules: ditto

- Update bokehutils dependency. [Per Unneberg]

  Fix #22

- Reuse matplotlib >= 1.4.0. [Per Unneberg]

- Test matplotlib 1.4.0. [Per Unneberg]

- Readd matplotlib dependency. [Per Unneberg]

- Revert conf and remove matplotlib dependency. [Per Unneberg]

- Mock: alternative take. [Per Unneberg]

- Mock matplotlib for documentation build. [Per Unneberg]

  Attempt to fix #21

- Fix errors in readme. [Per Unneberg]

- Misc fixes. [Per Unneberg]

  - remove test doc files
  - readd load_sml_config

0.1-alpha.6 (2015-08-25)
------------------------

- Add release notes. [Per Unneberg]

- Delete travis test branch. [Per Unneberg]

- Remove superfluous test line. [Per Unneberg]

- Add pytest to setup. [Per Unneberg]

- Make py.test default test. [Per Unneberg]

- Use update_config. [Per Unneberg]

  Minor fix.

  Fixes #18 and #19

- Use update_config throughout. [Per Unneberg]

  - for multiple update_config calls, need specific config_default dicts
    for each call (otherwise config points to a reference)

- Make unique test file names. [Per Unneberg]

  py.test best practices recommends unique file names


scripts/make2snakemake.py
#!/usr/bin/env python
# File: make2snakemake.py
# Created: Sat Nov  1 13:09:47 2014
# Copyright (C) 2014 by Per Unneberg
#
# Author: Per Unneberg
#

"""
Convert biomake rules to snakemake
"""
import sys
import re
import os

rules = []
config = {}
targets = []
states = [None, "rule", "config"]

header = """
# -*- snakemake -*-
import os
from snakemakelib.utils import update_config, sml_rules_path

# Start by including the general snakefile
include: os.path.join(sml_rules_path(), 'base_settings.rules')
"""

if __name__ == "__main__":
    rule = {}
    if len(sys.argv) != 2:
        print("Usage: ", sys.argv[0], " makefile")
    else:
        state = None
        with open(sys.argv[1], 'r') as fp:
            for l in fp.readlines():
                line = l.rstrip()
                if re.search("^[^ ]+:", line) and not line.startswith("print") and not line.startswith("#"):
                    state = "rule"
                elif line.startswith("ifndef"):
                    state = "config"
                elif line.startswith("endif"):
                    state = None
                elif line.startswith("%"):
                    state = "rule"
                elif line.startswith("\t"):
                    pass
                elif not line:
                    if state == "rule":
                        rules.append(rule)
                    state = None
                elif line.startswith("#"):
                    state = None
                elif re.search("([A-Z_a-z0-9]+)=(.*)", line):
                    if state is None:
                        state = "vardef"
                else:
                    pass
                # Add to config and rule
                if state == "config":
                    if not line.startswith("ifndef") and not line.startswith("ifneq") and not line.startswith("ifeq"):
                        m = re.match("([^=]+)=(.*)", line)
                        (k,v) = m.groups()
                        config[k] = v
                elif state == "rule":
                    print (line)
                    if re.search("^%", line):
                        (output, inputs)  = [x.replace("%", "{prefix}") for x in line.split(":")]
                        rule = {'input': inputs,
                                'output' : output,
                                'shell' : "",
                                'name' : ""}
                    elif re.search("(^[^ %]+):", line):
                        m = re.search("(^[^ %]+):(.*)", line)
                        rule = {'input': "",
                                'output' : "",
                                'shell' : "",
                                'name' : m.group(1)}
                        if m.group(2):
                            rule["input"] = m.group(2)
                    else:
                        rule["shell"] += "{shell}".format(shell=line.lstrip())
                elif state=="vardef":
                    targets.append(line)

        
        print(header)
        # Format config
        print ("config_default = {leftbrace} \n\t\"{section}\" : {leftbrace}".format(leftbrace="{", section=os.path.basename(sys.argv[1].replace("Makefile.", ""))))
        for (section,opt) in config.items():
            print ("\t\t\"{section}\" : \"{opt}\",".format(section=section, opt=opt))
        print ("\t\t{rightbrace},\n\t{rightbrace},\n{rightbrace}".format(rightbrace="}"))

        print ("\nconfig = update_config(config, config_default)\n")

        # print targets
        for tgt in targets:
            print (tgt)
        # Format rules
        i = 0
        for rule in rules:
            i = i + 1
            if rule["name"]:
                print ("rule {name}:".format(name=rule["name"]))
            else:
                print ("rule rule_{cnt}:".format(cnt=i))
            for k in ["input", "output", "shell"]:
                if rule[k]:
                    print ("\t{k}: \"{v}\"".format(k=k, v=rule[k]))


